sdk/helper/keysutil/encrypted_key_storage.go

```diff
@@ -307,3 +321,169 @@ func (s *encryptedKeyStorage) encryptPath(path string) (string, error) {
 
 	return encPath, nil
 }
+
+// eksTransaction differs from other implementations of transactions in that
+// we want to explicitly avoid poisoning the lru above.
+type eksTransaction struct {
+	t  *transactionalEncryptedKeyStorage
+	tx logical.Transaction
+}
+
+var _ logical.Transaction = &eksTransaction{}
+
+func (t *transactionalEncryptedKeyStorage) BeginReadOnlyTx(ctx context.Context) (logical.Transaction, error) {
+	tx, err := t.encryptedKeyStorage.s.(logical.TransactionalStorage).BeginReadOnlyTx(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &eksTransaction{
+		t:  t,
+		tx: tx,
+	}, nil
+}
+
+func (t *transactionalEncryptedKeyStorage) BeginTx(ctx context.Context) (logical.Transaction, error) {
+	tx, err := t.encryptedKeyStorage.s.(logical.TransactionalStorage).BeginTx(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &eksTransaction{
+		t:  t,
+		tx: tx,
+	}, nil
+}
+
+// List implements the logical.Storage List method, and decrypts all the items
+// in a path prefix. This can only operate on full folder structures so the
+// prefix should end in a "/".
+func (e *eksTransaction) List(ctx context.Context, prefix string) ([]string, error) {
+	return e.ListPage(ctx, prefix, "", -1)
+}
+
+// ListPage implements the logical.Storage List method, and decrypts a subset
```

Comment:
Is reusing the LRU bad here? I don't actually think it is: all it does is help with decrypting keys in the underlying list, which means we can drop this implementation and reuse the underlying one instead.